Update microgrid client

Marenz opened the pull request:
Blocked by:
* ~The release of base-client v0.11~
* ~The release of microgrid client~

ela-kotulska-frequenz commented on benchmarks/power_distribution/power_distributor.py:
```diff
@@ -12,7 +12,8 @@
 
 from frequenz.channels import Broadcast
-from frequenz.client.microgrid import Component, ComponentCategory, ComponentId
+from frequenz.client.microgrid import Component, ComponentCategory
+from frequenz.client.microgrid.id import ComponentId
```
This is breaking change! (I just added this import `from frequenz.client.microgrid import ComponentId` to all depending repositories)
And we plan to remove ComponentId form `from frequenz.client.microgrid` anyway, soon.

Please consider reverting this change from `frequenz-client-microgrid-python` and leave `ComponentId` in `frequenz.client.microgrid`.

llucax replied:
We'll need to import it from the microgrid client and release it before we can do that in the SDK. It should not take too long, so unless this PR needs to be merged urgently, I think it might be worth going directly for the [IDs in `client-common`](https://github.com/frequenz-floss/frequenz-client-common-python/pull/77).

BTW, I added this breaking change to the ugrid client release notes, it was missing there.

llucax commented on src/frequenz/sdk/microgrid/_power_distributing/_component_managers/_battery_manager.py:
```diff
@@ -11,14 +11,15 @@
 from datetime import timedelta
 
 from frequenz.channels import LatestValueCache, Receiver, Sender
+from frequenz.client.base.streaming import filter_stream_events
```
I guess this needs updating anyways, as `filter_stream_events` is also not happening. So I will stop reviewing here until the "dependencies" are "stable".

Marenz committed:
Update microgrid client to v0.9.0

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

llucax commented on pyproject.toml:
```diff
@@ -170,14 +171,22 @@ max-attributes = 12
 # Please see this issue for more details:
 # https://github.com/frequenz-floss/frequenz-sdk-python/issues/982
-addopts = "-W=all -Werror -Wdefault::DeprecationWarning -Wdefault::PendingDeprecationWarning -Wdefault::pytest.PytestUnraisableExceptionWarning -vv"
+addopts = "-Wdefault::pytest.PytestUnraisableExceptionWarning -vv"
```
```suggestion
addopts = "-vv"
```

All warnings config should be moved to the `filterwarnings` option.

llucax commented on pyproject.toml:
```diff
@@ -170,14 +171,22 @@ max-attributes = 12
 markers = [
   "integration: integration tests (deselect with '-m \"not integration\"')",
 ]
+filterwarnings = [
+  "error",
+  "once::DeprecationWarning",
+  "once::PendingDeprecationWarning",
+  # We use a raw string (single quote) to avoid the need to escape special
+  # chars as this is a regex
+  'ignore:Protobuf gencode version .*exactly one major version older.*:UserWarning',
+]
```
I wonder if you added this just for consistency or if you are also getting protobuf warnings at this level :scream:

Marenz replied:
I am getting warnings indeed.

llucax replied:
Damn, I didn't know they were so intrusive...

llucax commented on pyproject.toml:
```diff
@@ -170,14 +171,22 @@ max-attributes = 12
+filterwarnings = [
+  "error",
+  "once::DeprecationWarning",
+  "once::PendingDeprecationWarning",
```
```suggestion
  "once::PendingDeprecationWarning",
  "default::pytest.PytestUnraisableExceptionWarning",
```

llucax commented on pyproject.toml:
```diff
@@ -163,21 +164,30 @@ disable = [
+  # We use a raw string (single quote) to avoid the need to escape special
+  # chars as this is a regex
+  'ignore:Protobuf gencode version .*exactly one major version older.*:UserWarning',
+]
```
Nitpick, but for extra clarity I would group the lines without comments first, and then put the lines with comments:

```suggestion
filterwarnings = [
  "error",
  "once::DeprecationWarning",
  "once::PendingDeprecationWarning",
  # We need to ignore PytestUnraisableExceptionWarning because we have some
  # cleanup issues, and sometimes exceptions are raised in __del__ methods, for
  # example when trying to do async stuff and the event loop is already closed.
  # This could even be caused by the GC, some time after the test function that
  # has the issue was completed.
  # Please see this issue for more details:
  # https://github.com/frequenz-floss/frequenz-sdk-python/issues/982
  "default::pytest.PytestUnraisableExceptionWarning",
  # We use a raw string (single quote) to avoid the need to escape special
  # chars as this is a regex
  'ignore:Protobuf gencode version .*exactly one major version older.*:UserWarning',
]
```

llucax commented:
[Test with nox (amd64, ubuntu-24.04, 3.13, pytest_min)](https://github.com/frequenz-floss/frequenz-sdk-python/actions/runs/15589617127/job/43904982238#logs) is consistently timing out due to some infinite loop while cleaning up for tests (tests finish, but the reporting of warnings, etc. is never ending) :worried:

llucax commented:
Not sure if it might be related or a similar issue to the `time-machine` update issues we had and @sahas-subramanian-frequenz fixed.

<details>
<summary>These are the logs after the tests finish</summary>

```
================= 425 passed, 4 warnings in 119.58s (0:01:59) ==================
Task was destroyed but it is pending!
task: <Task pending name='LatestValueCache«BatteryManager«0x7fd88c486350»:battery«CID9»»' coro=<LatestValueCache._run() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_latest_value_cache.py:107> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='LatestValueCache«BatteryManager«0x7fd88c486350»:inverter«CID8»»' coro=<LatestValueCache._run() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_latest_value_cache.py:107> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='LatestValueCache«BatteryManager«0x7fd88c486350»:battery«CID29»»' coro=<LatestValueCache._run() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_latest_value_cache.py:107> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='LatestValueCache«BatteryManager«0x7fd88c486350»:inverter«CID28»»' coro=<LatestValueCache._run() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_latest_value_cache.py:107> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='LatestValueCache«BatteryManager«0x7fd88c486350»:battery«CID19»»' coro=<LatestValueCache._run() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_latest_value_cache.py:107> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='LatestValueCache«BatteryManager«0x7fd88c486350»:inverter«CID18»»' coro=<LatestValueCache._run() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_latest_value_cache.py:107> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='LatestValueCache«BatteryManager«0x7fd88c773200»:inverter«CID48»»' coro=<LatestValueCache._run() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_latest_value_cache.py:107> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='#17' coro=<MetricFetcher.fetch_next() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/timeseries/formula_engine/_formula_steps.py:514> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='#7' coro=<MetricFetcher.fetch_next() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/timeseries/formula_engine/_formula_steps.py:514> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='#7' coro=<MetricFetcher.fetch_next() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/timeseries/formula_engine/_formula_steps.py:514> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='Task-18511' coro=<sleep() running at /opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/tasks.py:718> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='Task-18512' coro=<Event.wait() running at /opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/locks.py:213> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='Task-18750' coro=<Event.wait() running at /opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/locks.py:213> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='Task-18916' coro=<Event.wait() running at /opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/locks.py:213> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='Task-18948' coro=<Event.wait() running at /opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/locks.py:213> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='Task-18958' coro=<Event.wait() running at /opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/locks.py:213> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='Task-18960' coro=<Event.wait() running at /opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/locks.py:213> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Exception ignored in: <function BackgroundService.__del__ at 0x7fd88fb92980>
Traceback (most recent call last):
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_background_service.py", line 276, in __del__
    self.cancel("{self!r} was deleted")
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_actor.py", line 148, in cancel
    return super().cancel(msg)
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_background_service.py", line 171, in cancel
    task.cancel(msg)
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 827, in call_soon
    self._check_closed()
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 550, in _check_closed
Error:     raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function BackgroundService.__del__ at 0x7fd88fb92980>
Traceback (most recent call last):
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_background_service.py", line 276, in __del__
    self.cancel("{self!r} was deleted")
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_background_service.py", line 171, in cancel
    task.cancel(msg)
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 827, in call_soon
    self._check_closed()
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 550, in _check_closed
Error:     raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function BackgroundService.__del__ at 0x7fd88fb92980>
Traceback (most recent call last):
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_background_service.py", line 276, in __del__
    self.cancel("{self!r} was deleted")
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_background_service.py", line 171, in cancel
    task.cancel(msg)
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 827, in call_soon
    self._check_closed()
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 550, in _check_closed
Error:     raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function BackgroundService.__del__ at 0x7fd88fb92980>
Traceback (most recent call last):
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_background_service.py", line 276, in __del__
    self.cancel("{self!r} was deleted")
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_background_service.py", line 171, in cancel
    task.cancel(msg)
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 827, in call_soon
    self._check_closed()
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 550, in _check_closed
Error:     raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Task was destroyed but it is pending!
task: <Task pending name='0' coro=<Receiver.__anext__() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_receiver.py:201> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[_wait.<locals>._on_completion() at /opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/tasks.py:521]>
Task was destroyed but it is pending!
task: <Task pending name='1' coro=<Receiver.__anext__() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_receiver.py:201> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[_wait.<locals>._on_completion() at /opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/tasks.py:521]>
Task was destroyed but it is pending!
task: <Task pending name='2' coro=<Receiver.__anext__() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_receiver.py:201> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[_wait.<locals>._on_completion() at /opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/tasks.py:521]>
Task was destroyed but it is pending!
task: <Task pending name='Task-24021' coro=<ComponentPoolStatusTracker._run() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/microgrid/_power_distributing/_component_pool_status_tracker.py:117> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task cancelling name='Task-24022' coro=<Actor._run_loop() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_actor.py:102> wait_for=<Future cancelled>>
Exception ignored in: <function BackgroundService.__del__ at 0x7fd88fb92980>
Traceback (most recent call last):
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_background_service.py", line 276, in __del__
    self.cancel("{self!r} was deleted")
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_actor.py", line 148, in cancel
    return super().cancel(msg)
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_background_service.py", line 171, in cancel
    task.cancel(msg)
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 827, in call_soon
    self._check_closed()
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 550, in _check_closed
Error:     raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Task was destroyed but it is pending!
task: <Task cancelling name='Task-24023' coro=<Actor._run_loop() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_actor.py:102> wait_for=<Future cancelled>>
Task was destroyed but it is pending!
task: <Task cancelling name='Task-24025' coro=<BatteryStatusTracker._run() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/microgrid/_power_distributing/_component_status/_battery_status_tracker.py:267> wait_for=<Future cancelled>>
Task was destroyed but it is pending!
task: <Task cancelling name='Task-24026' coro=<BatteryStatusTracker._run() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/microgrid/_power_distributing/_component_status/_battery_status_tracker.py:267> wait_for=<Future cancelled>>
Task was destroyed but it is pending!
task: <Task cancelling name='Task-24027' coro=<BatteryStatusTracker._run() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/microgrid/_power_distributing/_component_status/_battery_status_tracker.py:267> wait_for=<Future cancelled>>
Task was destroyed but it is pending!
task: <Task pending name='LatestValueCache«BatteryManager«0x7fd88c7359d0»:battery«CID29»»' coro=<LatestValueCache._run() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_latest_value_cache.py:107> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='LatestValueCache«BatteryManager«0x7fd88c7359d0»:battery«CID19»»' coro=<LatestValueCache._run() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_latest_value_cache.py:107> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='LatestValueCache«BatteryManager«0x7fd88c7359d0»:inverter«CID48»»' coro=<LatestValueCache._run() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_latest_value_cache.py:107> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='LatestValueCache«BatteryManager«0x7fd88c7359d0»:inverter«CID38»»' coro=<LatestValueCache._run() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_latest_value_cache.py:107> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Exception ignored in: <coroutine object ComponentPoolStatusTracker._run at 0x7fd88c07a940>
Traceback (most recent call last):
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/microgrid/_power_distributing/_component_pool_status_tracker.py", line 112, in _run
    async with contextlib.AsyncExitStack() as stack:
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/contextlib.py", line 768, in __aexit__
    raise exc
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/contextlib.py", line 751, in __aexit__
    cb_suppress = await cb(*exc_details)
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_background_service.py", line 226, in __aexit__
    await self.stop()
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_background_service.py", line 190, in stop
    await self.wait()
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_background_service.py", line 241, in wait
    done, pending = await asyncio.wait(self._tasks)
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/tasks.py", line 450, in wait
    loop = events.get_running_loop()
RuntimeError: no running event loop
Actor PowerDistributingActor[140568046327552]: Raised a BaseException.
Traceback (most recent call last):
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_actor.py", line 102, in _run_loop
    await self._run()
GeneratorExit
Actor PowerManagingActor[140568044333424]: Raised an unhandled exception during stop.
Traceback (most recent call last):
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_select.py", line 438, in select
    done, pending = await asyncio.wait(
                    ^^^^^^^^^^^^^^^^^^^
        pending, return_when=asyncio.FIRST_COMPLETED
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
GeneratorExit

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/actor/_actor.py", line 102, in _run_loop
    await self._run()
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/microgrid/_power_managing/_power_managing_actor.py", line 224, in _run
    async for selected in select(
    ...<61 lines>...
            self._algorithm.drop_old_proposals(asyncio.get_event_loop().time())
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_select.py", line 470, in select
    await _stop_pending_tasks(pending)
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_select.py", line 484, in _stop_pending_tasks
    task.cancel()
    ~~~~~~~~~~~^^
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 827, in call_soon
    self._check_closed()
    ~~~~~~~~~~~~~~~~~~^^
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 550, in _check_closed
Error:     raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Task was destroyed but it is pending!
task: <Task cancelling name='8785502476025' coro=<_Receiver.ready() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_broadcast.py:475> wait_for=<Future cancelled>>
BatteryStatusTracker crashed with error: Event loop is closed
Traceback (most recent call last):
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_select.py", line 438, in select
    done, pending = await asyncio.wait(
                    ^^^^^^^^^^^^^^^^^^^
        pending, return_when=asyncio.FIRST_COMPLETED
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
GeneratorExit

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/microgrid/_power_distributing/_component_status/_battery_status_tracker.py", line 267, in _run
    async for selected in select(
    ...<47 lines>...
            )
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_select.py", line 470, in select
    await _stop_pending_tasks(pending)
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_select.py", line 484, in _stop_pending_tasks
    task.cancel()
    ~~~~~~~~~~~^^
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 827, in call_soon
    self._check_closed()
    ~~~~~~~~~~~~~~~~~~^^
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 550, in _check_closed
Error:     raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Task was destroyed but it is pending!
task: <Task cancelling name='8785502476039' coro=<_Receiver.ready() running at /home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_broadcast.py:475> wait_for=<Future cancelled>>
Exception ignored in: <coroutine object _Receiver.ready at 0x7fd88c0e0e40>
Traceback (most recent call last):
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_broadcast.py", line 475, in ready
    await self._channel._recv_cv.wait()
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/locks.py", line 300, in wait
    self._notify(1)
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/locks.py", line 341, in _notify
    fut.set_result(False)
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 827, in call_soon
    self._check_closed()
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py", line 550, in _check_closed
Error:     raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
BatteryStatusTracker crashed with error: no running event loop
Traceback (most recent call last):
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/microgrid/_power_distributing/_component_status/_battery_status_tracker.py", line 267, in _run
    async for selected in select(
    ...<47 lines>...
            )
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_select.py", line 435, in select
    pending.add(asyncio.create_task(recv.ready(), name=name))
                ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/tasks.py", line 407, in create_task
    loop = events.get_running_loop()
RuntimeError: no running event loop
/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/microgrid/_power_distributing/_component_status/_battery_status_tracker.py:318: RuntimeWarning: coroutine '_Receiver.ready' was never awaited
  _logger.exception("BatteryStatusTracker crashed with error: %s", err)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
BatteryStatusTracker crashed with error: no running event loop
Traceback (most recent call last):
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/microgrid/_power_distributing/_component_status/_battery_status_tracker.py", line 267, in _run
    async for selected in select(
    ...<47 lines>...
            )
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_select.py", line 435, in select
    pending.add(asyncio.create_task(recv.ready(), name=name))
                ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/tasks.py", line 407, in create_task
    loop = events.get_running_loop()
RuntimeError: no running event loop
BatteryStatusTracker crashed with error: no running event loop
Traceback (most recent call last):
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/microgrid/_power_distributing/_component_status/_battery_status_tracker.py", line 267, in _run
    async for selected in select(
    ...<47 lines>...
            )
  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/.nox/pytest_min/lib/python3.13/site-packages/frequenz/channels/_select.py", line 435, in select
    pending.add(asyncio.create_task(recv.ready(), name=name))
                ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/tasks.py", line 407, in create_task
    loop = events.get_running_loop()
RuntimeError: no running event loop
```

(keeps looping on the last runtime error in "  File "/home/runner/work/frequenz-sdk-python/frequenz-sdk-python/src/frequenz/sdk/microgrid/_power_distributing/_component_status/_battery_status_tracker.py", line 267").

</details>

Marenz committed:
BackgroundService: Remove __del__ method

We have no guarantee that the event loop is still running at the time
the __del__ method is called. In fact we do see errors in the
in the test cleanup phase exactly because of this.

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

shsms commented on src/frequenz/sdk/actor/_background_service.py:
```diff
@@ -268,13 +268,6 @@ def __await__(self) -> collections.abc.Generator[None, None, None]:
         """
         return self.wait().__await__()
 
-    def __del__(self) -> None:
-        """Destroy this instance.
-
-        Cancel all running tasks spawned by this background service.
-        """
-        self.cancel("{self!r} was deleted")
-
```
I think the fix is to check if the event loop is running before calling cancel, because when the loop is there, we still want to cancel and cleanup.

Marenz replied:
I thought about that, but I don't think we gain much from that tbh. Cleanup should not rely on `del`
